services/optimiser/optimiser.py

```diff
@@ -23,209 +27,165 @@ def __init__(self, session: orm.session, request_id: int, debug: bool):
-        if result is not None:
-            for asset_request_index, asset_request in enumerate(result['contrib']):
-                for volunteer_index, volunteer in enumerate(asset_request):
-                    for role_index, assigned in enumerate(volunteer):
-                        if assigned:
-                            role_domain = self.calculator.get_role_by_index(role_index)
-                            user_domain = self.calculator.get_volunteer_by_index(volunteer_index)
-                            asset_request_domain = self.calculator.get_asset_request_by_index(asset_request_index)
-                            print(
-                                f'Volunteer {user_domain.email} should be assigned to role {role_domain.code} on asset request {asset_request_domain.id}')
-                            asset_request_obj = persist[asset_request_index]
-                            role_map = asset_request_obj[role_domain.id]
-                            role_map['assigned'].append(user_domain.id)
-
-        # Now, actually persist it!
-        for asset_request_index, _ in enumerate(persist):
-            asset_request = self.calculator.get_asset_request_by_index(asset_request_index)
-            asset_request_roles = persist[asset_request_index]
-            for role_id in asset_request_roles:
-                for assign_count in range(asset_request_roles[role_id]['count']):
-                    if assign_count < len(asset_request_roles[role_id]['assigned']):
-                        ar = AssetRequestVolunteer(user_id=asset_request_roles[role_id]['assigned'][assign_count],
-                                                   vehicle_id=asset_request.id, role_id=role_id, status='pending')
-                    else:
-                        ar = AssetRequestVolunteer(user_id=None, vehicle_id=asset_request.id, role_id=role_id)
-                    session.add(ar)
+        shift_request_id = self.calculator.request_id
+
+        try:
+            # Process the MiniZinc result by iterating over shifts, roles, and volunteers
+            for shift_index, shift_assignments in enumerate(result["possible_assignment"]):  # Iterate over shifts
+                shift = self.calculator._shifts_[shift_index]  # Directly access the shift by index
+                for role_index, role_assignments in enumerate(shift_assignments):  # Iterate over roles in the shift
+                    for volunteer_index, is_assigned in enumerate(role_assignments):  # Iterate over volunteers
+                        if is_assigned:  # If a volunteer is assigned to a role for this shift
+                            user = self.calculator.get_volunteer_by_index(volunteer_index)
+                            role = self.calculator.get_role_by_index(role_index)
+
+                            # Create a ShiftRequestVolunteer entry with status PENDING
+                            shift_volunteer = ShiftRequestVolunteer(
```

Comment:
Please do this through a repository

Reply:
Hi Ben, I believe I've implemented the required changes now; please do give feedback if I have implemented anything wrong.